docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,324 @@
+
+**Story 4: Define drain order & exclude Pods from drain without management cluster access**
+
+As a workload cluster admin/user, I want to be able to define drain order of Pods and also exclude Pods from drain 
+(similar to Story 1-3, just without management cluster access).
+
+**Story 5: Apply drain configuration to a set of Machines / Clusters**
+
+As a cluster admin/user, I would like to be able to configure the drain behavior for all Machines of a management 
+cluster or for a subset of these Machines without having to configure each Machine individually. I would also like 
+to be able to change the drain configuration without having to modify all Machine objects.
+
+### Goals
+
+* Allow users to customize how/if Pods are evicted:
+  * Allow to define the order in which Pods are evicted
+  * Allow to define if Pods should be excluded from drain
+* Use the current drain default rules as a fallback if neither `MachineDrainRule` CRs nor drain annotations
+  apply to a Pod (this also ensures compatibility with current behavior)
+* Stop waiting for volume detachment for Pods that are excluded from drain
+
+### Non-Goals
+
+* Change the drain behavior for DaemonSet and static Pods (we’ll continue to skip draining for both)
+
+### Future work
+
+* Align with the evolving landscape for Node drain in Kubernetes as soon as the new features are available, 
+  stable and can be used for the Kubernetes versions Cluster API supports (e.g. [KEP-4212](https://github.com/kubernetes/enhancements/issues/4212), [KEP-4563](https://github.com/kubernetes/enhancements/issues/4563)).
+
+  Because of the wide range of workload cluster Kubernetes versions supported in Cluster API (n-5), it will take 
+  significant time until we can use new Kubernetes features in all supported versions. So an implementation in
+  Cluster API is the only option to improve drain anytime soon (even for the oldest supported Kubernetes versions).
+
+  Also, please note that while we would like to align with Kubernetes and leverage its new features as soon as 
+  possible, there will always be a delta to be addressed in Cluster API due to the fact that Cluster API is operated 
+  from the management cluster, while drain is done in workload clusters.
+
+  Additionally, Cluster API should take care of many clusters, while Kubernetes features are by design scoped to a single cluster.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### `MachineDrainRule` CRD
+
+We propose to introduce a new `MachineDrainRule` namespace-scoped CRD. The goal of this CRD is to allow management 
+cluster admins/users to define a drain rule (a specific drain behavior) for a set of Pods.
+
+Additionally, drain rules can be restricted to a specific set of Machines, e.g. targeting only controlplane Machines, 
+Machines of specific Clusters, Linux Machines, etc.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDrainRule
```

Comment:
MachineDrainRule vs DrainRule?

Reply:
We initially had NodeDrainRule :) We moved to MachineDrainRule because it nicely correlates with the Machine CRD